test/functional/keypool-restore.py

```diff
@@ -0,0 +1,157 @@
+        self.num_nodes = 2
+        self.extra_args = [['-usehd=0'], ['-usehd=1', '-keypool=100', '-keypoolemin=20']]
+
+    def run_test(self):
+        tmpdir = self.options.tmpdir
+
+        self.nodes[0].generate(101)
+
+        self.log.info("Make backups of unencrypted and encrypted wallet")
+        # stop and backup original wallet (only keypool has been initialized)
+        self.stop_node(1)
+        shutil.copyfile(tmpdir + "/node1/regtest/wallet.dat", tmpdir + "/hd.bak")
+
+        # start again and encrypt wallet
+        self.nodes[1] = self.start_node(1, self.options.tmpdir, self.extra_args[1])
+        self.nodes[1].encryptwallet('test')
+        self.bitcoind_processes[1].wait(timeout=BITCOIND_PROC_WAIT_TIMEOUT)
+        # node will be stopped during encryption, now do a backup
+        shutil.copyfile(tmpdir + "/node1/regtest/wallet.dat", tmpdir + "/hd.enc.bak")
+
+        self.log.info("Generate addresses and send funds to addresses in the keypool and extended keypool")
+        # start the node with encrypted wallet, get address in new pool at pos 50 (over the gap limit)
+        self.nodes[1] = self.start_node(1, self.options.tmpdir, self.extra_args[1])
+        for _ in range(90):
+            addr_enc_oldpool = self.nodes[1].getnewaddress()
+
+        # now make sure we retrive an address in the extended pool
+        self.nodes[1].walletpassphrase("test", 10)
+        for _ in range(20):
+            addr_enc_extpool = self.nodes[1].getnewaddress()
+
+        # stop and load initial backup of the unencrypted wallet
+        self.stop_node(1)
+        shutil.copyfile(tmpdir + "/hd.bak", tmpdir + "/node1/regtest/wallet.dat")
+        self.nodes[1] = self.start_node(1, self.options.tmpdir, self.extra_args[1])
+        connect_nodes_bi(self.nodes, 0, 1)
+
+        addr = self.nodes[1].getnewaddress()
+        assert_equal(self.nodes[1].validateaddress(addr)['hdkeypath'], "m/0'/0'/1'")
+
+        rawch = self.nodes[1].getrawchangeaddress()
+        assert_equal(self.nodes[1].validateaddress(rawch)['hdkeypath'], "m/0'/1'/0'")
+
+        # Send funds (one in the main keypool over the gap limit, the other in the extended pool space)
+        self.nodes[0].sendtoaddress(addr_enc_oldpool, 10)
+        self.nodes[0].generate(1)
+        stop_height = self.nodes[0].getblockchaininfo()['blocks']
+        self.nodes[0].sendtoaddress(addr_enc_extpool, 5)
+        self.nodes[0].generate(1)
+
+        # self.nodes[0].sendtoaddress(addr, 1)
+
+        # n0addr = self.nodes[0].getnewaddress()
+        # txdata = self.nodes[0].createrawtransaction([], {rawch: 2.0, n0addr: 3.0})
+
+        # txdata_f = self.nodes[0].fundrawtransaction(txdata)
+        # txdata_s = self.nodes[0].signrawtransaction(txdata_f['hex'])
+        # self.nodes[0].sendrawtransaction(txdata_s['hex'])
+
+        # self.sync_all()
+        # self.nodes[0].generate(1)
+        self.sync_all()
+
+        assert_equal(self.nodes[1].getbalance(), 15)
```

Comment:
On test failure happening this line, I think the reason is that encryptwallet generates an entirely new HD master key, so `addr_enc_oldpool` and `addr_enc_extpool` come from a new HD master key which is not part of "hd.bak"

Restoring "hd.enc.bak" instead of "hd.bak" makes this check pass: https://github.com/ryanofsky/bitcoin/commit/93349ad5d915ee152a1a84920f18cbe0cab8b036